core/shared/platform/include/platform_api_extension.h

```diff
@@ -555,6 +622,376 @@ os_socket_set_recv_timeout(bh_socket_t socket, uint64 timeout_us);
+                                uint32_t imr_interface);
+
+/**
+ * Drop membership of a group
+ *
+ * @param socket the socket to set timeout
+ * @param mreq the group multicast address and local interface address
+ *
+ * @return 0 if success, -1 otherwise
+ */
+int
+os_socket_set_ip_drop_membership(bh_socket_t socket, uint32_t imr_multiaddr,
+                                 uint32_t imr_interface);
+
+/**
+ * Set the current time-to-live field that is
+ * used in every packet sent from this socket.
+ * @param socket the socket to set the flag
+ * @param ttl_s time to live (seconds)
+ *
+ * @return 0 if success, -1 otherwise
+ */
+int
+os_socket_set_ip_ttl(bh_socket_t socket, uint8_t ttl_s);
+
+/**
+ * Retrieve the current time-to-live field that is
+ * used in every packet sent from this socket.
+ * @param socket the socket to set the flag
+ * @param ttl_s time to live (seconds)
+ *
+ * @return 0 if success, -1 otherwise
+ */
+int
+os_socket_get_ip_ttl(bh_socket_t socket, uint8_t *ttl_s);
+
+/**
+ * Set the time-to-live value of outgoing multicast
+ * packets for this socket
+ * @param socket the socket to set the flag
+ * @param ttl_s time to live (seconds)
+ *
+ * @return 0 if success, -1 otherwise
+ */
+int
+os_socket_set_ip_multicast_ttl(bh_socket_t socket, uint8_t ttl_s);
+
+/**
+ * Read the time-to-live value of outgoing multicast
+ * packets for this socket
+ * @param socket the socket to set the flag
+ * @param ttl_s time to live (seconds)
+ *
+ * @return 0 if success, -1 otherwise
+ */
+int
+os_socket_get_ip_multicast_ttl(bh_socket_t socket, uint8_t *ttl_s);
+
+/**
+ * Restrict to sending and receiving IPv6 packets only
+ *
+ * @param socket the socket to set
+ * @param enable 1 to enable or 0 to disable
+ *
+ * @return 0 if success, -1 otherwise
+ */
+int
+os_socket_set_ipv6_only(bh_socket_t socket, bool enable);
+
+/**
+ * Get whether only sending and receiving IPv6 packets
+ *
+ * @param socket the socket to check
+ * @param enabled 1 for enabled or 0 for disabled
+ *
+ * @return 0 if success, -1 otherwise
+ */
+int
+os_socket_get_ipv6_only(bh_socket_t socket, bool *enabled);
+
+/**
+ * Set enable or disable IPv6 multicast loopback.
+ *
+ * @param socket the socket to set
+ * @param enable 1 to enable or 0 to disable
+ *
+ * @return 0 if success, -1 otherwise
+ */
+int
+os_socket_set_ipv6_multicast_loop(bh_socket_t socket, bool enable);
+
+/**
+ * Get enable or disable IPv6 multicast loopback.
+ *
+ * @param socket the socket to check
+ * @param enabled 1 for enabled or 0 for disabled
+ *
+ * @return 0 if success, -1 otherwise
+ */
+int
+os_socket_get_ipv6_multicast_loop(bh_socket_t socket, bool *enabled);
+
+/**
+ * Join IPv6 multicast group
+ *
+ * @param socket the socket to set timeout
+ * @param mreq the group multicast address and local interface address
+ *
+ * @return 0 if success, -1 otherwise
+ */
+int
+os_socket_set_ipv6_join_group(bh_socket_t socket, uint16_t imr_multiaddr[8],
```

Comment:
So, referring to my comment above - we could possibly merge this (and other methods that are for IPv4 and IPv6) into one method (rather than having 2) and pass `is_ipv6` flag - unless you think it's better to keep them separate?